docs/src/modules/ROOT/pages/enterprise-edition/enterprise-edition.adoc

```diff
@@ -759,5 +760,191 @@ the host is likely to hang or freeze,
+
+var a = factory.forEach(Shift.class)
+               .filter(predicate);
+
+var b = factory.forEach(Shift.class)
+               .filter(predicate);
+----
+
+Whereas these building blocks are not functionality equivalent:
+
+[source,java,options="nowrap"]
+----
+Predicate<Shift> predicate1 = shift -> shift.getEmployee().getName().equals("Ann");
+Predicate<Shift> predicate2 = shift -> shift.getEmployee().getName().equals("Bob");
+
+// Different parents
+var a = factory.forEach(Shift.class)
+               .filter(predicate2);
+
+var b = factory.forEach(Shift.class)
+               .filter(predicate1)
+               .filter(predicate2);
+
+// Different operations
+var a = factory.forEach(Shift.class)
+               .ifExists(Employee.class);
+
+var b = factory.forEach(Shift.class)
+               .ifNotExists(Employee.class);
+
+// Different inputs
+var a = factory.forEach(Shift.class)
+               .filter(predicate1);
+
+var b = factory.forEach(Shift.class)
+               .filter(predicate2);
+----
+
+Counterintuitively, the building blocks produced by these (seemly) identical methods are not necessarily functionality equivalent:
+
+[source,java,options="nowrap"]
+----
+UniConstraintStream<Shift> a(ConstraintFactory constraintFactory) {
+    return factory.forEach(Shift.class)
+                  .filter(shift -> shift.getEmployee().getName().equals("Ann"));
+}
+
+UniConstraintStream<Shift> b(ConstraintFactory constraintFactory) {
+    return factory.forEach(Shift.class)
+                  .filter(shift -> shift.getEmployee().getName().equals("Ann"));
+}
+----
+
+The Java Virtual Machine is free to (and often does) create different instances of functionality equivalent lambdas.
+This severely limits the effectiveness of node sharing, since the only way to know two lambdas are equal is to compare their references.
+
+When automatic node sharing is used, the `ConstraintProvider` class is transformed so all lambdas are accessed via a static final field.
+So a class that looks like this:
+
+[source,java,options="nowrap"]
+----
+public class MyConstraintProvider implements ConstraintProvider {
+
+    public Constraint[] defineConstraints(ConstraintFactory constraintFactory) {
+        return new Constraint[] {
+            a(constraintFactory),
+            b(constraintFactory)
+        };
+    }
+
+    Constraint a(ConstraintFactory constraintFactory) {
+        return factory.forEach(Shift.class)
+                      .filter(shift -> shift.getEmployee().getName().equals("Ann"))
+                      .penalize(SimpleScore.ONE)
+                      .asConstraint("a");
+    }
+
+    Constraint b(ConstraintFactory constraintFactory) {
+        return factory.forEach(Shift.class)
+                      .filter(shift -> shift.getEmployee().getName().equals("Ann"))
+                      .penalize(SimpleScore.ONE)
+                      .asConstraint("b");
+    }
+}
+----
+
+is transformed to a class that looks like this:
```

Comment:
When automatic node sharing is enabled, the class will be transformed to look like this: